### DISCUSSION TITLE
Terminal on a notcurses plane

### DISCUSSION THREAD

**Discussion by tomek-szczesny** [dankamongmen/notcurses#1831] [Ideas]
Hey there!
I've decided to implement a multi-server ssh client, something like clusterssh, but with TUI. I got a grasp on libssh, that's easy, but the problem of displaying a few terminals in a single terminal is far more complicated than I hoped.
As far as I know, there are only two programs on the face of Earth that catch and reinterpret escape codes of underlying processes, which are `screen` and `tmux`. Both implement terminals within themselves, as far as I can tell.

As I was reading htp-notcurses.pdf I stumbled upon a mention of `ncprocess` widget that was somehow meant to address this problem. Unfortunately, this widget has been mentioned only once, and even Google thinks no such thing exists.

But anyway, notcurses is a WIP, and very promising indeed - so I think it would greatly benefit from a widget that emulates a terminal. If I'm about to implement such functionality anyway, could as well be a part of a library.

One way to achieve that is using [libvterm](http://www.leonerd.org.uk/code/libvterm/), I believe, a library that parses escape codes and calls callbacks. Anyway, I'm not sure if you consider using an external library kosher.

So, to wrap this up, I wanted to ask what is the current state of affairs. Am I missing something that would ease my task, what happened to `ncprocess`, is a "terminal widget" a thing you'd like to see in this library?

Thanks a lot!

**Comment by joseluis** [comment]
Hi there!

You can take a look to these issues to see what's been discussed so far:
- https://github.com/dankamongmen/notcurses/issues/487
- https://github.com/dankamongmen/notcurses/issues/1332

I personally think this would be a very good addition to the library...

**Reply by dankamongmen** [reply]
yep, @joseluis linked the exact two issues i intended to. the delay on these has been threefold:

* i started this project with essentially no idea of how terminals worked, and it has been a tremendous learning experience. when i started in November of 2019, i thought CSI stood for "crime scenes investigation" and didn't know what an alternate screen was. the stuff you're talking about requires serious knowledge of terminal receive side, not client send side, and i just didn't have the knowledge to embark on such an effort. i know a great deal more now =].
* i personally do not use any kind of terminal multiplexer, and am thus ignorant of how people *want* them to work and the real ideal cases
* i wanted to get the client side to a state of completion before expanding upon it. i knew that this journey would start with sane Unicode support, and that it would end with multiplanar bitmaps, and that i had no real idea what would live in the middle, and it would probably take about two years of my life and effort. that all turned out to be accurate.

but so now we're just about done with all of that, and i have grown (imho) into a fairly knowledgeable terminal developer. the time for projects like those of which you speak draws nigh. let's ride.

[![do the thing, maverick](http://img.youtube.com/vi/siwpn14IE7E/0.jpg)](http://www.youtube.com/watch?v=siwpn14IE7E "highway to the daaaaanger zone")

**Reply by tomek-szczesny** [reply]
Hey, that's awesome!

I've recently been actively digging through Interwebs to understand the topic better and I think the problem goes down to this.
We get a stream of text, such as data coming over SSH link or stdout of pty. We want to print it on a plane / in a widget, and the said stream is the only input to this widget.
Escape sequences must be isolated from the text. Text gets printed and escape codes are applied accordingly.
I see there were already mentions of libvterm, and I see it as a parser that does just that - separates text from control data.
BTW libvterm either had some sort of schism or is just too obvious of a name for a library. The one that I linked appears to be THE libvterm.

The beauty of this task is that it can be easily broken down into smaller chunks and worked on by more people. The baseline minimum requires reacting to cursor-moving escape codes and scrolling the content on each newline. Then we may get fancy and support colors, scrollable history buffer, and stuff like that,

Just to make it clear, I'm not a professional coder, and not that handy with C either, so I don't think I stand a chance in producing anything worth a pull request. However I can prepare a working mock-up that could be translated into Pedantic C. Would that be of any help?

**Reply by dankamongmen** [reply]
yep i completely agree. as i've always seen it, there are the following tasks:

* [ ] ideally we'd prevent the subprocess from opening the controlling tty using e.g. `/dev/tty`, as the results can't be good (though this might mean some programs refuse to start, or misbehave, hrmm). but if the subprocess can talk directly to the tty, it destroys our authority and control.
* [ ] control the subprocess's concept of its own geometry. if we're running it in some 20x20 subregion of the 140x70 visible area, we don't want the subprocess thinking it has 140x70 available. how do we do this? we can export `LINES` and `COLUMNS`, but really we need to override the `TIOCGWINSZ` `ioctl(2)` IMHO.
* [ ] we need the ability to launch a subprocess and connect it to a plane. we pretty much have this in `ncsubprocess`
* [ ] as you said, extract control sequences. i'd like to do this as part of a general markup plugin ideally, see #1554. in my mind, handling tty terminfo input oughtn't be too different from handling HTML or markdown -- you're taking in-band signaling, and turning it into terminfo
* [ ] we would *probably* (maybe?) need introduce our own terminfo entry, and deal with all that
* [ ] we need the ability to scroll planes and the actual window content. the former has long existed through `ncplane_set_scrolling()`. the latter just became possible with #1823, but it's very new functionality, and is likely to be complex and volatile

i'm sure there's more, but that's what i can see thus far.

so a lot of this is going to be pretty deep in the core, which i'm not sure is the easiest place for someone to join up. but maybe not? i definitely welcome PRs, but have not yet had anyone really submitting into the core, and am not sure how that'll go. if you want to move on this, i'd recommend filing issues on what you intend to do before beginning work, so that i can get in on the design with you. i'd try to keep the PRs small and self-contained early on, until we build up a working rapport. i'll become quickly less guarded after you prove yourself a bit, but right now i'm still spending 100% of my time on developing the core out, and anything that overwhelms me with novelty or danger or being too much Unlike What I Would Do is just going to be frustrating for everyone, because i tend to respond to things either within a few hours or about three months later.

sorry for all the wariness, just we're talking about a pretty complicated and substantial piece of work, lower down in the stack than anything anyone but me has done, and i'm prickly and protective lol.

**Reply by dankamongmen** [reply]
there's also a very real question of whether or not this belongs in Notcurses, or whether it ought be a library that makes use of Notcurses. is it something that can be built atop Notcurses? it doesn't need to *encapsulate* Notcurses, and almost certainly should not -- i'm thinking more that you could have a library that provides `ncterm_create()` accepting an `ncplane*` like other widgets. i'm not saying it should be that way, just wondering whether it could/ought.

**Reply by tomek-szczesny** [reply]
Heh, just in time when I brewed my morning coffee. ;)

I think we're talking about two separate things now:
- A widget that shows text and interprets terminal's control signals (string stream input or whatever). 
- An actual terminal living on a noncurses plane, which may execute anything on its own.

Personally I'm after the dumber version 1. I already know how to collect data, just need means for displaying them. This is indeed pretty close to #1554 and might as well just be a part of it.

I agree building a whole terminal inside notcurses, although would be really cool, should not be anywhere near the top of the priority list. After all, it's a TUI library and I feel it should primarily provide looks and not logic. I'd say to let the target user deal with ptys and stuff.

And finally, sure I understand you care about you baby project and why you police the way it is expanding. I'm not here to cause any trouble, but then again thanks for basic guidelines how to contribute.

If we agree that it's just a matter of making #1554 a reality, I could try implementing such a parser. Or if someone else starts implementing it, I could do the tedious job of pairing html/markdown/VT100 control with notcurses functions.
Like I said, I'm going to do that anyway, so would be cool to make it a part of something more useful to the general public.

**Reply by dankamongmen** [reply]
oh you're not talking about taking this output from a subprocess? that's much less complicated, indeed. how are you getting the output, then? what's emitting VT codes that's not a subprocess?

**Reply by dankamongmen** [reply]
asked another way, i understand that this is effectively a streaming function taking UTF8 to `ncplane_*()` output functions. that's all well and good, but there are presumably cursor movement escapes in there. how do you intend to limit said cursor movements to ones which make sense for arbitrary planes?

**Reply by tomek-szczesny** [reply]
Like I said on the very beginning, the source of the text with VT codes is the SSH host. I implemented SSH connection using libssh instead of firing up the external ssh client. I have a file-like object that I can read data from, and I want to simply put it in notcurses plane/widget.

Cursor movement is my major concern, because without it even basic maintenance task like "apt upgrade" will look like shit (progress bar).
The SSH host is aware of the terminal size and should not try to move cursor beyond these bounds. Of course, if the plane's size changes, I'll have to report it to ssh host, but that's outside of the scope of markup parsing feature.

**Reply by dankamongmen** [reply]
> Like I said on the very beginning, the source of the text with VT codes is the SSH host.

sorry, i missed this!

> Cursor movement is my major concern, because without it even basic maintenance task like "apt upgrade" will look like shit (progress bar). The SSH host is aware of the terminal size and should not try to move cursor beyond these bounds. Of course, if the plane's size changes, I'll have to report it to ssh host, but that's outside of the scope of markup parsing feature.

yep, so this is what i see as the biggest open problem. you say that the ssh process is "aware of the terminal size". it's almost certainly not, actually (why would it need to be? it just transports bytes). the things that are aware of the size are the remote programs you run that care about terminal size, which are most likely running a `TIOCGWINSZ` `ioctl(2)` on the pty allocated by the ssh process. IMHO, we want to shim that somehow. i'm not sure of any of this -- i'm looking into it now. but this is what we need control if we're to control emitted geometry.

i don't see any way to generally map mismatched geometries. it's possible for very limited output, or if you're willing to allow the virtual "screen" to be truncated, but otherwise i don't see how. and a complicated program which changes its output drastically based on available size just needs to know the true output space available.

if you're planning on just putting it through without using a subregion, you shouldn't need to worry about it much?

i would recommend:

* accepting a plane rather than creating it yourself, for sure
* finding the geometry of that plane, and
* truncating any output which would lie outside of the plane

then we want to combine that with information about the real virtual screen size, then we want to launch processes such that they are constrained to the desired virtual screen size. but for now, it sounds like virtual size == real size, so things ought work out naturally.

**Reply by tomek-szczesny** [reply]
The SSH host generally is aware of the client terminal size (just checked with `stty -a` on a remote machine).

[This](https://api.libssh.org/stable/group__libssh__channel.html#gaf5d55c90f3d98c583df23d21905c1127) is how I notify the ssh host about the terminal size. I'm not 100% sure about this, but this should trigger the signal on the remote process to resize accordingly (`SIGWINCH`?) I may also inspect the library code and see how exactly this is handled.

If the remote side is running a program that reacts to such an event (vim, htop), it will surely redraw the whole screen, so that's not a problem. I don't know if Bash and others are smart enough to capture that and act on it. IF not, perhaps there is another way to force the SSH host to resend the contents of a pty.

But then again, I don't feel it's much of a concern from the notcurses perspective.

**Reply by dankamongmen** [reply]
it knows it because it runs the `TIOCGWINSZ` `ioctl(2)` on the pty created by ssh, as i mentioned:

```c
[killermike](1) $ strace stty -a 2>&1 | grep TIOC
ioctl(1, TIOCGWINSZ, 0x7ffc01e81df0)    = -1 ENOTTY (Inappropriate ioctl for device)
ioctl(0, TIOCGWINSZ, {ws_row=70, ws_col=107, ws_xpixel=1070, ws_ypixel=1400}) = 0
[killermike](0) $ 
```

indeed, it doesn't work if you use `ssh stty -a` without `-t` to force pty allocation:

```c
[schwarzgerat](0) $ ssh killermike strace stty -a 2>&1 | grep TIOC
[schwarzgerat](1) $ ssh killermike -t strace stty -a 2>&1 | grep TIOC
ioctl(1, TIOCGWINSZ, {ws_row=70, ws_col=107, ws_xpixel=1070, ws_ypixel=1400}) = 0
ioctl(0, TIOCGWINSZ, {ws_row=70, ws_col=107, ws_xpixel=1070, ws_ypixel=1400}) = 0
[schwarzgerat](0) $ 
```

**Reply by dankamongmen** [reply]
here it is without -t and without redirection:

```
[schwarzgerat](1) $ ssh killermike stty -a 
stty: 'standard input': Inappropriate ioctl for device
[schwarzgerat](1) $ 
```

`stty -a` doesn't work in this case because there is no pty to interrogate with the `ioctl(2)`.

**Reply by tomek-szczesny** [reply]
Luckily, libssh DOES allocate pty. :) That's why the function in question is called `ssh_channel_change_pty_size()`.

**Reply by dankamongmen** [reply]
yep, that works for your case! great!

my whole point is just that *in general* i'm going to want to figure out how to control the `ioctl(2)`s results, so that we can force these programs to conform to a desired geometry, so that i can do something like

```
n = ncplane_create(40x40 plane)
nsub = ncsubprocess_create(40x40, "ssh foo blah")
while(read(nsub, buffer)){
  ncplane_puthtml(buffer);
}
```

or maybe even

```
n = ncplane_create(40x40 plane)
nsub = ncsubprocess_create(40x40, "ssh foo blah")
ncfilter_create(puthtml, nsub, n);
```

**Reply by tomek-szczesny** [reply]
I think I'll just go ahead and run some experiments, and cry for help when stuff won't work as I imagined. ;)

**Reply by dankamongmen** [reply]
but as i've said, you don't need any of that infrastructure for what you propose to do here.

**Reply by dankamongmen** [reply]
sounds great!

**Reply by tomek-szczesny** [reply]
If my experiment works, it may be viable to inspire ncprocess off the libssh's pty management methods. But like I said before, I don't really think notcurses should be expected to provide ncprocess anytime soon. Might as well expect libssh to format a nice output, which sounds daft to me.

**Reply by dankamongmen** [reply]
yep, completely agreed that the current work doesn't need it. just wanted to think about it.

**Reply by tomek-szczesny** [reply]
> yep, completely agreed that the current work doesn't need it. just wanted to think about it.

I like where's your head at. Thinking about stuff is my lifetime habit. Glad to meet people that do the same.

**Reply by dankamongmen** [reply]
and i'm glad to have smart Poles wielding English at the level of Józef Teodor Konrad Korzeniowski coming out of nowhere to help out on my little project! if your code is as dexterous as your angielski, you're likely to be a big help indeed =]. Mogę jeść szkło, nie boli mnie to!

**Reply by tomek-szczesny** [reply]
Wouldn't count on that, I'm just a simple electronics engineer in a lazy space industry, not a coder. ;) But I think you may hope for at least a working prototype.

**Comment by tomek-szczesny** [comment]
Time for an off topic ramble (as if most of this discussion was on topic)

So I dug into the library and successfully used it. Created an nc_plane, put text in there and all was well. It seems to drop consistent portions of repeated ssh sessions, and it appears that VT spells present in the read buffer make `ncplane_putnstr()` go apeshit. Also, a finite length of a buffer breaks some escape sequences, which certainly doesn't help.

That's one problem of ncplane_putvt() - it should be assumed that a user will just throw buffer contents right at it, and doing so may break escape sequences (or multi-byte characters). The way I see it, either a user will have to deal with that, or virtual terminal functionality should belong to a dedicated widget, that will store interrupted endings and concatenate them in front of subsequent inputs.
ncplane_putvt() technically could store these stubs in itself and keep track of them for all ncplanes but that's just awfully messy.

Now from the user's perspective, I'm sorry to say that but the project would surely benefit from more verbose and organized manual. There's a book, which ought to be outdated pretty soon. There are some md files in doc/, which are empty aside of the basics, that says basically the same what the book says. There is USAGE.md which is the most helpful, but very hard to navigate. And finally one Wiki page which is an interesting read but doesn't say much about usage.
I think that although you @dankamongmen said you're not a fan of example-based tutoring, this is precisely what most people want. Probably because a certain generation is severely addicted to instant rewards and can't give two shits about 100+ pages read before anything interesting appears on the screen. Or maybe some people just want a recipe from a cookbook to make stuff work before polishing it.
Once I get a good grasp of the library and deal with virtual terminal thing, I may allow myself to write a short tutorial based on my experiences.

**Reply by tomek-szczesny** [reply]
I take it back - libvterm has NO documentation whatsoever and I claim it unusable. Docs are not included in any package, and the original web location has been wiped out. 
No wonder why tmux and screen have their own takes on parsing VT spam (aside from the fact these projects are much older than libvterm).

So I was thinking which approach is the best to develop a parser. Is it yacc/bison code generator, [some general parsing library](https://github.com/orangeduck/mpc), or (frankly my first idea) a state machine done by hand. The latter is probably the closest to the original implementation in the physical terminal, and the closest to my actual skills. ANSI screams state machine.
Nevertheless, I promised at least a working prototype, so for now I might as well pick anything.
Parsing HTML, bbcode or markdown has been dropped from my list for now, because all of these are inherently stateless (as long as user inputs complete chunks of content). Terminals, on the other hand, have at least a few bytes of state - there are commands that store and recall cursor position, for example. This calls for VT widget rather than VT parsing function, albeit heavily based on ncplane.

Anyway I'm starting to think this will be a truly unique feature of notcurses. If you think about it, this will make notcurses a sole viable library to implement a true TUI desktop environment. This is fucking exciting.
Screw implementing a better tmux/screen alternative, go full DE! A text-based i3 clone would take the FLOSS community by storm.

**Reply by dankamongmen** [reply]
> I take it back - libvterm has NO documentation whatsoever and I claim it unusable. Docs are not included in any package, and the original web location has been wiped out.

libvterm did not overwhelm with its quality when i looked into it, it is true. it dimmed my eagerness to work on this stuff.

> (frankly my first idea) a state machine done by hand

i definitely don't think yacc is applicable, since this ought be a regular language and thus at the level of flex.

i don't know about you, but i find flex a bit ungainly. i am similarly inclined towards hand-consed state machines.

on the other hand, we have exactly such a state machine in `src/lib/input.c`, and it does not bring me great delight. i have a sinking feeling that the first CVE against this project will originate in that code.

> Anyway I'm starting to think this will be a truly unique feature of notcurses. If you think about it, this will make notcurses a sole viable library to implement a true TUI desktop environment. This is fucking exciting.
> Screw implementing a better tmux/screen alternative, go full DE! A text-based i3 clone would take the FLOSS community by storm.

heh there's a dude working on that (kinda): https://github.com/o-sdn-o/VTM 

i'd like the interface of any such state machine to be streaming -- i.e., i can call it with 1 <= N bytes of input and it will run through them, and maintain enough context to pick up where it left off. this ought include any necessary UTF-8 character and EGC division.

the only thing that really makes sense in terms of an interface IMHO is something like:

`ssize_t ncplane_putvt(struct ncplane*n, struct ncvtctx* vtctx, const char* buf, size_t s);`

returning the number of bytes interpreted on success, and the negative number of bytes interpreted on failure, failing as soon as there is any kind of problem. as things are read, it is translating them into operations on the ncplane. failures can happen either due to a bad lex or a bad output operation.

maybe we need some kind of recovery. i'm unsure.

is this similar to what you were thinking?

**Reply by tomek-szczesny** [reply]
> i definitely don't think yacc is applicable, since this ought be a regular language and thus at the level of flex.
> i don't know about you, but i find flex a bit ungainly. i am similarly inclined towards hand-consed state machines.

Actually I learned about all these existing solutions yesterday, when I did some research. I didn't like them either, but for some reason I felt that doing stuff by hand is generally discouraged these days. I have no problem coding all that myself and definitely will give it some thought to make it look and feel clean.

> on the other hand, we have exactly such a state machine in `src/lib/input.c`

Great, then it may be reused even internally. What is this currently used for?

> heh there's a dude working on that (kinda): https://github.com/o-sdn-o/VTM

Drop-in replacements tend to get more attention. That's why maybe an i3-compatible DE would be seen as something more than yet another peculiarity. I know Sway got hugely successful because it just works with i3 config files and all - it's currently one of the better recognized Wayland compositors out there. i3 is biased towards keyboard operation by design.
But then again, that's just a thing to think about in spare time, far beyond our current capabilities.

> i'd like the interface of any such state machine to be streaming -- i.e., i can call it with 1 <= N bytes of input and it will run through them, and maintain enough context to pick up where it left off. this ought include any necessary UTF-8 character and EGC division.

Same here, I'd prefer to just yank data from SSH channel and put it right into `ncplane_putvt()` without any further complications. There is a need to store some context data and I wasn't sure how to approach this issue, but I like your idea. That's even better, because vtctx may be optional (== NULL), if a user is confident this will never be necessary for their application (for example print the output of a command only once).

> returning the number of bytes interpreted on success, and the negative number of bytes interpreted on failure, failing as soon as there is any kind of problem. as things are read, it is translating them into operations on the ncplane. failures can happen either due to a bad lex or a bad output operation.

Yes, precisely what I was planning to do - take a look at each byte one by one and feed them into the state machine. I just hope the efficiency of such approach will be acceptable - honestly no idea. If I did that with TTLs in 1975 I can tell it would be state of the art. We shall see how well it translates to machine code, I guess.

I'm not worried about error recovery for now - as far as I'm concerned, no mainstream terminals or standards defined how to handle erroneous input. The only source of errorrs for now are negative returns from underlying ncplane operation calls.

Thanks for your time - now I've got a really solid idea what I'm aiming for. I'll create a demo with a state machine sufficient to handle `toilet --gay dupa` and work from there. Sounds okay to you?

**Reply by dankamongmen** [reply]
> Actually I learned about all these existing solutions yesterday, when I did some research. I didn't like them either, but for some reason I felt that doing stuff by hand is generally discouraged these days. I have no problem coding all that myself and definitely will give it some thought to make it look and feel clean.

there are no beautiful alternatives here.

i concur that the tightness of the VT language makes it pretty amenable to hand-coded FSA.


> > on the other hand, we have exactly such a state machine in `src/lib/input.c`
> Great, then it may be reused even internally. What is this currently used for?

nooo. it processes replies to our initial burst of query strings. it's the other direction from where you want to be.

everything else sounds good!

the one thing i'd maybe recommend is that you have your different states be different functions, and not different cases in a switch like mine in `input.c`, so that we can check that all possible paths have a sensible destination (though this is basically equivalent to error recovery). but don't take that as a hard requirement. gigantic state machines are definitely difficult to review IMHO but that's probably independent of scope fracture [shrug].

one thing is that i think we'll need a terminfo database corresponding to this, and it should be built up alongside the lexer. it ought not be cribbed from an existing definition. i was reminded of this thinking about how we're going to test this. which, we'll need to have unit tests around this for sure. take a look at `src/tests` and `notcurses-tester`.

ummmm bonus points if you wrap a fuzzer around it.

hack on!

**Reply by tomek-szczesny** [reply]
Good, you're slowly getting the idea of my capabilities. That's a bunch of tips I needed, thanks. 
I haven't thought of making function-based state machine but the idea is sound to me. Should be much easier to manage errors, and especially sudden end of string. 

> gigantic state machines are definitely difficult to review IMHO

This is why I want to start up small, handle a few color codes and cursor moshing around the pty. As I said, I'll try to manage `toilet`'s colorful output first, but later I shall be challenged with `htop`'s ncurses-invoked cursor movement.

I'll try to keep an up-to-date list of supported spells, so it's easier to update terminfo entry. Gotta learn how to do it though, I assume you have no prior experience with that either, but it shouldn't be that hard, I reckon.

I have no idea about software tests, but I'll take a look at that later on. Fuzz test should be easy - just input random data and see if it crashes, right?
